File: ekocpp/src/clients/multiplexed_stream_client_base.cpp
#include "multiplexed_stream_client_base.hpp"

#include <array>

#include <asio/experimental/awaitable_operators.hpp>
#include <spdlog/spdlog.h>

#include "../data_transfer_objects/binary_frame.hpp"
#include "../exceptions/exceptions.hpp"

using namespace asio::experimental::awaitable_operators;

template <typename SocketType>
MultiplexedStreamClientBase<SocketType>::MultiplexedStreamClientBase(
    asio::any_io_executor executor, std::chrono::milliseconds timeout, std::chrono::milliseconds heartbeat_period,
    int max_retries, std::chrono::milliseconds retry_delay,
    std::chrono::milliseconds staleness_sweep_period, std::chrono::milliseconds staleness_warning_threshold
) : ClientBase(max_retries, retry_delay),
    executor_(executor),
    timeout_(timeout),
    heartbeat_period_(heartbeat_period),
    staleness_sweep_period_(staleness_sweep_period),
    staleness_warning_threshold_(staleness_warning_threshold),
    heartbeat_timer_(executor_),
    staleness_sweep_timer_(executor_),
    heartbeat_done_(executor_, 1),
    staleness_sweep_done_(executor_, 1),
    write_permit_(executor_, 1),
    connect_permit_(executor_, 1) {
    write_permit_.try_send(std::error_code{});   // one permit available immediately
    connect_permit_.try_send(std::error_code{}); // one permit available immediately
}

template <typename SocketType>
void MultiplexedStreamClientBase<SocketType>::start() {
    asio::co_spawn(executor_, run_heartbeat_loop<SocketType>(this->weak_from_this()), asio::detached);
    asio::co_spawn(executor_, run_staleness_sweep_loop<SocketType>(this->weak_from_this()), asio::detached);
}

template <typename SocketType>
asio::awaitable<void> MultiplexedStreamClientBase<SocketType>::stop() {
    stopping_.store(true);
    heartbeat_timer_.cancel();       // wakes both loops immediately instead of waiting out their periods
    staleness_sweep_timer_.cancel();
    co_await heartbeat_done_.async_receive(asio::use_awaitable);
    co_await staleness_sweep_done_.async_receive(asio::use_awaitable);
}

template <typename SocketType>
asio::awaitable<void> MultiplexedStreamClientBase<SocketType>::ensure_connected() {
    co_await connect_permit_.async_receive(asio::use_awaitable);
    struct PermitGuard {
        asio::experimental::concurrent_channel<void(std::error_code)>& permit;
        ~PermitGuard() { permit.try_send(std::error_code{}); }
    } guard{connect_permit_};

    if (!connected_) {
        socket_    = co_await open_connection();
        connected_ = true;
        asio::co_spawn(executor_, run_reader_loop<SocketType>(this->weak_from_this()), asio::detached);
    }
}

template <typename SocketType>
void MultiplexedStreamClientBase<SocketType>::fail_all_outstanding(std::error_code ec) {
    std::lock_guard<std::mutex> lock(demux_mutex_);
    for (auto& [key, entry] : demux_map_) {
        entry.channel->try_send(ec, std::vector<uint8_t>{});
    }
    demux_map_.clear();
}

// Mirrors PersistentStreamClientBase::do_heartbeat's intent (a bare ping/pong
// round trip, answered by the server's transport layer alone, never reaching
// its router) but rides the same demux as every other caller -- it cannot
// perform its own independent async_read the way the Persisted version does,
// since the reader loop is now the socket's sole reader. See
// multiplexed_stream_client.md's "Reconnection" section for why this is
// forced correctness, not a style choice.
template <typename SocketType>
asio::awaitable<void> MultiplexedStreamClientBase<SocketType>::do_heartbeat() {
    try {
        co_await ensure_connected();
    } catch (const std::system_error&) {
        co_return; // connection refused
    }

    const SpanKey                 ping_key = SpanKey::generate();
    auto                          channel  = std::make_shared<DemuxChannel>(executor_, 1);
    {
        std::lock_guard<std::mutex> lock(demux_mutex_);
        demux_map_[ping_key] = DemuxEntry{channel, std::chrono::steady_clock::now(), false};
    }

    try {
        const auto ping = pack_ping_frame(ping_key);

        co_await write_permit_.async_receive(asio::use_awaitable);
        {
            struct PermitGuard {
                asio::experimental::concurrent_channel<void(std::error_code)>& permit;
                ~PermitGuard() { permit.try_send(std::error_code{}); }
            } guard{write_permit_};
            co_await asio::async_write(*socket_, asio::buffer(ping), asio::use_awaitable);
        }

        asio::steady_timer timer(co_await asio::this_coro::executor, timeout_);
        auto result = co_await (
            channel->async_receive(asio::use_awaitable) || timer.async_wait(asio::use_awaitable)
        );

        {
            std::lock_guard<std::mutex> lock(demux_mutex_);
            demux_map_.erase(ping_key);
        }

        if (result.index() == 1) {
            connected_ = false; // timed out waiting for the pong
            co_return;
        }

        // A set error_code on the channel (fail_all_outstanding) is thrown as
        // a std::system_error by use_awaitable, not returned here -- caught
        // below, same as any other write/read failure.
        const auto& pong_frame = std::get<0>(result);
        if (pong_frame.size() < HEADER_LENGTH) {
            connected_ = false;
            co_return;
        }
        const ParsedHeader parsed = parse_header(pong_frame.data());
        if (!(parsed.flags & PING_FLAG)) {
            connected_ = false;
        }
    } catch (const std::system_error&) {
        connected_ = false;
    }
}

template <typename SocketType>
asio::awaitable<std::vector<uint8_t>> MultiplexedStreamClientBase<SocketType>::send_message_retry_loop(std::vector<uint8_t> request) {
    // co_await is not permitted inside a catch block -- the executor is
    // fetched once up front; see the matching comment in
    // persistent_stream_client_base.cpp.
    auto executor = co_await asio::this_coro::executor;

    const SpanKey span_key = parse_header(request.data()).span_key;
    int           retry_count = 0;

    while (retry_count < max_retries_) {
        bool should_retry = false;
        auto channel       = std::make_shared<DemuxChannel>(executor, 1);
        try {
            co_await ensure_connected();

            {
                std::lock_guard<std::mutex> lock(demux_mutex_);
                demux_map_[span_key] = DemuxEntry{channel, std::chrono::steady_clock::now(), false};
            }

            co_await write_permit_.async_receive(asio::use_awaitable);
            {
                struct PermitGuard {
                    asio::experimental::concurrent_channel<void(std::error_code)>& permit;
                    ~PermitGuard() { permit.try_send(std::error_code{}); }
                } guard{write_permit_};
                co_await asio::async_write(*socket_, asio::buffer(request), asio::use_awaitable);
            }

            // Deliberately NOT raced against a timer -- an ordinary send has
            // no per-request timeout. See multiplexed_stream_client.md's
            // "Per-request timeout" section: multiplexing means a slow
            // response no longer blocks anything else, and connection death
            // is already handled below via fail_all_outstanding. A
            // permanently-unanswered entry is a staleness-sweep warning, not
            // a retry trigger.
            //
            // A set error_code on the channel (fail_all_outstanding) is
            // thrown as a std::system_error by use_awaitable here, not
            // returned -- caught below, same as a failed write.
            std::vector<uint8_t> response_frame = co_await channel->async_receive(asio::use_awaitable);

            {
                std::lock_guard<std::mutex> lock(demux_mutex_);
                demux_map_.erase(span_key);
            }

            co_return response_frame;
        } catch (const std::system_error&) {
            {
                std::lock_guard<std::mutex> lock(demux_mutex_);
                demux_map_.erase(span_key);
            }
            connected_   = false;
            should_retry = true;
        }

        if (should_retry) {
            ++retry_count;
            if (retry_count >= max_retries_) {
                throw CommunicationsMaxRetriesReached("Maximum retries exceeded in sending request.");
            }
            asio::steady_timer retry_timer(executor, retry_delay_);
            co_await retry_timer.async_wait(asio::use_awaitable);
        }
    }
    throw CommunicationsMaxRetriesReached("Maximum retries exceeded in sending request.");
}

// Free function -- see the forward declaration's comment for why. Sole
// reader of the socket it's bound to (spawned fresh by ensure_connected()
// each time it reconnects -- see multiplexed_stream_client.md's
// "Reconnection" section for why one long-lived loop surviving reconnects
// was rejected in favour of this).
template <typename SocketType>
asio::awaitable<void> run_reader_loop(std::weak_ptr<MultiplexedStreamClientBase<SocketType>> weak_self) {
    for (;;) {
        auto self = weak_self.lock();
        if (!self) co_return; // owner is gone -- nothing left to read for.

        try {
            std::array<uint8_t, HEADER_LENGTH> header{};
            co_await asio::async_read(*self->socket_, asio::buffer(header), asio::use_awaitable);
            const ParsedHeader parsed = parse_header(header.data());

            std::vector<uint8_t> rest(parsed.total_len);
            if (parsed.total_len > 0) {
                co_await asio::async_read(*self->socket_, asio::buffer(rest), asio::use_awaitable);
            }

            std::vector<uint8_t> frame;
            frame.reserve(HEADER_LENGTH + rest.size());
            frame.insert(frame.end(), header.begin(), header.end());
            frame.insert(frame.end(), rest.begin(), rest.end());

            std::shared_ptr<typename MultiplexedStreamClientBase<SocketType>::DemuxChannel> channel;
            {
                std::lock_guard<std::mutex> lock(self->demux_mutex_);
                auto it = self->demux_map_.find(parsed.span_key);
                if (it != self->demux_map_.end()) {
                    channel = it->second.channel;
                    self->demux_map_.erase(it);
                }
            }
            if (channel) {
                channel->try_send(std::error_code{}, std::move(frame));
            }
            // else: orphan response (already timed out / stale) -- discarded.
        } catch (const std::system_error& e) {
            self->fail_all_outstanding(e.code());
            self->connected_ = false;
            co_return; // this loop's connection is dead -- the next ensure_connected() spawns a fresh one.
        }
    }
}

// Free function -- same lifetime discipline as PersistentStreamClientBase's
// run_heartbeat_loop.
template <typename SocketType>
asio::awaitable<void> run_heartbeat_loop(std::weak_ptr<MultiplexedStreamClientBase<SocketType>> weak_self) {
    for (;;) {
        auto self = weak_self.lock();
        if (!self) co_return;

        self->heartbeat_timer_.expires_after(self->heartbeat_period_);
        try {
            co_await self->heartbeat_timer_.async_wait(asio::use_awaitable);
        } catch (const std::system_error&) {
            break; // cancelled by stop() (or the timer's own destructor, if the owner forgot to call stop())
        }

        if (self->stopping_.load()) break;

        co_await self->do_heartbeat();
    }

    if (auto self = weak_self.lock()) {
        self->heartbeat_done_.try_send(std::error_code{});
    }
}

// Free function -- periodic sweep, structurally mirroring run_heartbeat_loop.
// Never removes entries, never retries, never touches the connection --
// purely operational visibility. See multiplexed_stream_client.md's
// "Per-request timeout" section.
template <typename SocketType>
asio::awaitable<void> run_staleness_sweep_loop(std::weak_ptr<MultiplexedStreamClientBase<SocketType>> weak_self) {
    for (;;) {
        auto self = weak_self.lock();
        if (!self) co_return;

        self->staleness_sweep_timer_.expires_after(self->staleness_sweep_period_);
        try {
            co_await self->staleness_sweep_timer_.async_wait(asio::use_awaitable);
        } catch (const std::system_error&) {
            break;
        }

        if (self->stopping_.load()) break;

        {
            std::lock_guard<std::mutex> lock(self->demux_mutex_);
            const auto now = std::chrono::steady_clock::now();
            for (auto& [key, entry] : self->demux_map_) {
                if (!entry.warned && (now - entry.registered_at) >= self->staleness_warning_threshold_) {
                    entry.warned = true;
                    spdlog::warn(
                        "MultiplexedStreamClientBase: request with span_key {} has been unanswered for over {}ms -- "
                        "connection is healthy, so this is likely a server-side bug that never wrote a response. "
                        "Not auto-retried; needs manual investigation.",
                        key.to_string(), self->staleness_warning_threshold_.count()
                    );
                }
            }
        }
    }

    if (auto self = weak_self.lock()) {
        self->staleness_sweep_done_.try_send(std::error_code{});
    }
}

template class MultiplexedStreamClientBase<asio::ip::tcp::socket>;
template class MultiplexedStreamClientBase<asio::local::stream_protocol::socket>;

template asio::awaitable<void> run_heartbeat_loop<asio::ip::tcp::socket>(
    std::weak_ptr<MultiplexedStreamClientBase<asio::ip::tcp::socket>>);
template asio::awaitable<void> run_heartbeat_loop<asio::local::stream_protocol::socket>(
    std::weak_ptr<MultiplexedStreamClientBase<asio::local::stream_protocol::socket>>);

template asio::awaitable<void> run_reader_loop<asio::ip::tcp::socket>(
    std::weak_ptr<MultiplexedStreamClientBase<asio::ip::tcp::socket>>);
template asio::awaitable<void> run_reader_loop<asio::local::stream_protocol::socket>(
    std::weak_ptr<MultiplexedStreamClientBase<asio::local::stream_protocol::socket>>);

template asio::awaitable<void> run_staleness_sweep_loop<asio::ip::tcp::socket>(
    std::weak_ptr<MultiplexedStreamClientBase<asio::ip::tcp::socket>>);
template asio::awaitable<void> run_staleness_sweep_loop<asio::local::stream_protocol::socket>(
    std::weak_ptr<MultiplexedStreamClientBase<asio::local::stream_protocol::socket>>);

File: ekosis/clients/multiplexed_stream_client_base.py
import time
import asyncio
import logging

from abc import ABC, abstractmethod
from typing import Dict, Tuple

from .client_base import ClientBase

from ..data_transfer_objects import HEADER_LENGTH, PING_FLAG, SpanKey, parse_header, pack_ping_frame

from ..exceptions import (
    CommunicationsNonRetryable,
    CommunicationsMaxRetriesReached,
)

log = logging.getLogger()


# --------------------------------------------------------------------------------
# One demux-map entry per request currently in flight. `future` is a single-use
# completion signal (asyncio's natural equivalent of ekocpp's single-slot
# concurrent_channel) -- the reader loop resolves it with the response frame,
# or fail_all_outstanding resolves it with an exception if the connection dies
# first. `warned` mirrors ekocpp's DemuxEntry: the staleness sweep logs once
# per entry, not every sweep.
class _DemuxEntry:
    __slots__ = ("future", "registered_at", "warned")

    def __init__(self, future: asyncio.Future):
        self.future       : asyncio.Future = future
        self.registered_at: float           = time.monotonic()
        self.warned       : bool            = False


# --------------------------------------------------------------------------------
# Python port of ekocpp's MultiplexedStreamClientBase -- full design history and
# rationale in ekocpp/documentation/multiplexed_stream_client.md. One connection,
# held open indefinitely (see that doc's "Lifecycle" section), but unlike
# PersistentStreamClientBase, many send_message() calls can be genuinely in
# flight at once: a write lock serialises only the write itself, and a
# background reader loop demuxes each incoming frame to its caller via the
# SpanKey already present in every response header.
#
# No explicit mutex around __demux_map -- unlike ekocpp, which genuinely runs
# the reader loop and writers across a thread pool, Python's asyncio is single-
# threaded and cooperative: dict mutations here are plain, non-await
# statements, so there's no interleaving window for a race the way there would
# be across real OS threads.
class MultiplexedStreamClientBase(ClientBase, ABC):
    def __init__(
        self,
        timeout                    : float = 5,
        heartbeat_period           : float = 60,
        max_retries                : int   = 3,
        retry_delay                : float = 0.1,
        staleness_sweep_period     : float = 30,
        staleness_warning_threshold: float = 60,
    ):
        super().__init__(max_retries, retry_delay)
        self.__timeout                    : float                     = timeout
        self.__heartbeat_period           : float                     = heartbeat_period
        self.__staleness_sweep_period      : float                     = staleness_sweep_period
        self.__staleness_warning_threshold : float                     = staleness_warning_threshold
        self.__connected                  : bool                      = False
        self.__stopping                   : bool                      = False
        self.__reader                     : asyncio.StreamReader      = None
        self.__writer                     : asyncio.StreamWriter      = None
        self.__write_lock                 : asyncio.Lock              = asyncio.Lock()
        self.__connect_lock               : asyncio.Lock              = asyncio.Lock()
        self.__demux_map                  : Dict[SpanKey, _DemuxEntry] = {}
        self.__reader_task                : asyncio.Task              = None
        self.__heartbeat_task             : asyncio.Task              = None
        self.__staleness_sweep_task       : asyncio.Task              = None

    # --------------------------------------------------------------------------------
    @abstractmethod
    async def open_connection(self) -> Tuple[asyncio.StreamReader, asyncio.StreamWriter]: # pragma: no cover
        pass

    # --------------------------------------------------------------------------------
    # Call once, after construction, to start the background heartbeat and
    # staleness-sweep loops. No shared_from_this() equivalent concern in
    # Python, but kept as an explicit call (not constructor-implicit) to mirror
    # ekocpp's lifecycle discipline -- callers of MultiplexedTCPClient/
    # MultiplexedUDSClient should start() and stop() them explicitly.
    def start(self):
        loop                         = asyncio.get_event_loop()
        self.__heartbeat_task       = loop.create_task(self.__run_heartbeat_loop())
        self.__staleness_sweep_task = loop.create_task(self.__run_staleness_sweep_loop())

    # --------------------------------------------------------------------------------
    # Await before releasing the last reference to this client, to confirm both
    # background loops (and the reader loop, if a connection is open) have
    # actually exited before teardown.
    async def stop(self):
        self.__stopping = True
        tasks = [t for t in (self.__heartbeat_task, self.__staleness_sweep_task, self.__reader_task) if t is not None]
        for task in tasks:
            task.cancel()
        for task in tasks:
            try:
                await task
            except asyncio.CancelledError:
                pass

    # --------------------------------------------------------------------------------
    # Guarded by __connect_lock -- NOT redundant with __write_lock. __write_lock
    # guards only the write itself, so multiple callers can race into
    # ensure_connected() concurrently on a fresh disconnect. Without its own
    # guard, that race can open two connections and spawn two reader loops for
    # what should be one. See multiplexed_stream_client.md.
    async def __ensure_connected(self):
        async with self.__connect_lock:
            if not self.__connected:
                self.__reader, self.__writer = await self.open_connection()
                self.__connected             = True
                self.__reader_task           = asyncio.get_event_loop().create_task(self.__run_reader_loop())

    # --------------------------------------------------------------------------------
    # Fails every outstanding demux entry at once (a dead connection takes out
    # everything in flight on it, not just whichever caller happened to notice
    # first) and clears the map.
    def __fail_all_outstanding(self, exc: Exception):
        for entry in self.__demux_map.values():
            if not entry.future.done():
                entry.future.set_exception(exc)
        self.__demux_map.clear()

    # --------------------------------------------------------------------------------
    # Mirrors PersistentStreamClientBase's do_heartbeat intent (a bare
    # ping/pong round trip, answered by the server's transport layer alone)
    # but rides the same demux as every other caller -- it cannot perform its
    # own independent read the way the Persisted version does, since the
    # reader loop is now the socket's sole reader.
    async def __do_heartbeat(self):
        try:
            await self.__ensure_connected()
        except (ConnectionRefusedError, OSError):
            return # connection refused

        ping_key                     = SpanKey.generate()
        future                       = asyncio.get_event_loop().create_future()
        self.__demux_map[ping_key]  = _DemuxEntry(future)

        try:
            async with self.__write_lock:
                self.__writer.write(pack_ping_frame(ping_key))
                await self.__writer.drain()

            pong_frame = await asyncio.wait_for(future, self.__timeout)

            _, _, _, flags = parse_header(pong_frame[:HEADER_LENGTH])
            if not (flags & PING_FLAG):
                self.__connected = False
        except (
            ConnectionResetError, ConnectionAbortedError, BrokenPipeError,
            asyncio.IncompleteReadError, OSError, asyncio.TimeoutError,
        ):
            self.__connected = False
        finally:
            self.__demux_map.pop(ping_key, None)

    # --------------------------------------------------------------------------------
    async def _send_message_retry_loop(self, request: bytes) -> bytes:
        span_key    = parse_header(request[:HEADER_LENGTH])[0]
        retry_count = 0

        while retry_count < self.max_retries:
            future                      = asyncio.get_event_loop().create_future()
            try:
                await self.__ensure_connected()
                self.__demux_map[span_key] = _DemuxEntry(future)

                async with self.__write_lock:
                    self.__writer.write(request)
                    await self.__writer.drain()

                # Deliberately NOT raced against a timeout -- an ordinary send
                # has no per-request timeout. See multiplexed_stream_client.md's
                # "Per-request timeout" section: multiplexing means a slow
                # response no longer blocks anything else, and connection
                # death is already handled by __fail_all_outstanding above. A
                # permanently-unanswered entry is a staleness-sweep warning,
                # not a retry trigger.
                response_frame = await future
                self.__demux_map.pop(span_key, None)
                return response_frame
            except (
                ConnectionResetError, ConnectionAbortedError, BrokenPipeError,
                asyncio.IncompleteReadError, OSError, ConnectionRefusedError,
            ) as e:
                self.__demux_map.pop(span_key, None)
                self.__connected = False
                retry_count     += 1
                if retry_count >= self.max_retries:
                    raise CommunicationsMaxRetriesReached() from e
                await asyncio.sleep(self.retry_delay)
            except Exception as e:
                self.__demux_map.pop(span_key, None)
                raise CommunicationsNonRetryable(f"{type(e)}: {str(e)}")

        raise CommunicationsMaxRetriesReached()

    # --------------------------------------------------------------------------------
    # Sole reader of the socket it's bound to -- spawned fresh by
    # __ensure_connected() each time it reconnects, rather than one long-lived
    # loop surviving reconnects (see multiplexed_stream_client.md's
    # "Reconnection" section for why).
    async def __run_reader_loop(self):
        try:
            while True:
                header                        = await self.__reader.readexactly(HEADER_LENGTH)
                span_key, _, total_len, _     = parse_header(header)
                rest                          = await self.__reader.readexactly(total_len) if total_len else b""
                frame                         = header + rest

                entry = self.__demux_map.pop(span_key, None)
                if entry is not None and not entry.future.done():
                    entry.future.set_result(frame)
                # else: orphan response (already timed out / stale) -- discarded.
        except asyncio.CancelledError:
            return
        except (asyncio.IncompleteReadError, ConnectionResetError, ConnectionAbortedError, BrokenPipeError, OSError) as e:
            self.__fail_all_outstanding(e)
            self.__connected = False

    # --------------------------------------------------------------------------------
    async def __run_heartbeat_loop(self):
        try:
            while True:
                await asyncio.sleep(self.__heartbeat_period)
                if self.__stopping:
                    return
                await self.__do_heartbeat()
        except asyncio.CancelledError:
            return

    # --------------------------------------------------------------------------------
    # Periodic sweep, structurally mirroring the heartbeat loop. Never removes
    # entries, never retries, never touches the connection -- purely
    # operational visibility. See multiplexed_stream_client.md's "Per-request
    # timeout" section: a permanently-unanswered request is a developer-visible
    # warning, not something the client silently heals.
    async def __run_staleness_sweep_loop(self):
        try:
            while True:
                await asyncio.sleep(self.__staleness_sweep_period)
                if self.__stopping:
                    return

                now = time.monotonic()
                for span_key, entry in self.__demux_map.items():
                    if not entry.warned and (now - entry.registered_at) >= self.__staleness_warning_threshold:
                        entry.warned = True
                        log.warning(
                            f"MultiplexedStreamClientBase: request with span_key {span_key} has been "
                            f"unanswered for over {self.__staleness_warning_threshold}s -- connection is "
                            f"healthy, so this is likely a server-side bug that never wrote a response. "
                            f"Not auto-retried; needs manual investigation."
                        )
        except asyncio.CancelledError:
            return
